subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/VariantMetadataRules.java

```diff
@@ -54,16 +56,38 @@ public class VariantMetadataRules {
     private CapabilitiesRules capabilitiesRules;
     private VariantFilesRules variantFilesRules;
 
+    private final AttributeContainerInternal baseAttributes;
+    private final Map<String, AttributeContainerInternal> variantAttributes = new ConcurrentHashMap<>();
```

Comment:
I was also curious. But, I hit a `ConcurrentModificationException` in a smoke test: 
https://builds.gradle.org/buildConfiguration/Gradle_Release_Check_Stage_PullRequestFeedback_Trigger/56268246?focusLine=0&hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

This exception can get thrown if the function passed to compute modifies the map, but also if `compute` is called concurrently. Since this code gets called during dependency resolution, and since dependency resolution runs within a build operation, this code can seemingly be called concurrently. Theoretically, this would be when separate configurations are being resolved which both reference dependencies modified by component metadata rules.